projects/driver_controls/src/center_console_io.c

```diff
@@ -0,0 +1,100 @@
+#include <stddef.h>
+#include <stdio.h>
+
+#include "center_console_io.h"
+#include "debounce.h"
+#include "driver_io.h"
+#include "gpio_expander.h"
+#include "gpio_it.h"
+#include "input_event.h"
+
+// Digital device identifiers
+typedef enum {
+  CENTER_CONSOLE_IO_DEVICE_POWER_SWITCH = 0,
+  CENTER_CONSOLE_IO_DEVICE_DIRECTION_SELECTOR,
+  CENTER_CONSOLE_IO_DEVICE_HEADLIGHT,
+  CENTER_CONSOLE_IO_DEVICE_HAZARD_LIGHT,
+  CENTER_CONSOLE_IO_DEVICE_BRAKING_REGEN_INC,
+  CENTER_CONSOLE_IO_DEVICE_BRAKING_REGEN_DEC
+} CenterConsoleIODevice;
+
+// Store the id of the device as well as the id of the event the device raises
+typedef struct CenterConsoleIOData {
+  CenterConsoleIODevice id;
+  InputEvent event;
+} CenterConsoleIOData;
+
+typedef struct CenterConsoleIOSettings {
+  GPIOAddress address;
+  InterruptEdge edge;
+} CenterConsoleIOSettings;
+
+// Index the objects using their respective pins
+static CenterConsoleIOData s_center_console_data[DRIVER_IO_NUM_ADDRESSES];
+
+static void prv_center_console_callback(const GPIOAddress *address, void *context) {
+  CenterConsoleIOData *data = (CenterConsoleIOData *)context;
+
+  GPIOState state = { 0 };
+  gpio_get_state(address, &state);
+  debounce(*address);
+
+  // Check the devices with multiple pins to ensure the correct event
+  uint16_t event_id = data->event;
+
+  if (data->id == CENTER_CONSOLE_IO_DEVICE_DIRECTION_SELECTOR && state == GPIO_STATE_HIGH) {
+    event_id = INPUT_EVENT_DIRECTION_SELECTOR_NEUTRAL;
+  }
+
+  event_raise(event_id, 0);
+}
+
+// Configure driver devices with their individual settings
+void center_console_io_init(void) {
+  // Initialize the static array with device information
+  s_center_console_data[DRIVER_IO_POWER_SWITCH_PIN] = ((CenterConsoleIOData){
```

Comment:
This seems to all be constant why not just store all this in a giant constant structure?

Reply:
Done